pkgs/applications/misc/rtfm/patches/make.patch

```diff
@@ -0,0 +1,30 @@
+--- a/Makefile	2023-07-17 17:18:28.000000000 +0300
++++ b/Makefile	2023-07-19 12:13:44.627168135 +0300
+@@ -4,8 +4,7 @@
+ all: configure .WAIT rtfm docsets
+ 
+ configure:
+-	shards install
+-	./bin/gi-crystal
++	gi-crystal
```

Comment:
Why you guys don't use the gi-crystal that the project depends on? Do you guys have a gi-crystal nix package? if so, why?

Reply:
I packed `gi-crystal` in `nixpkgs` when I packed the `collision`. You can see these PR's: https://github.com/NixOS/nixpkgs/pull/243914 
https://github.com/NixOS/nixpkgs/pull/242681
There is an explanation why we decided to do it this way.

Shortly, nix collects crystal dependencies in separated derivations, then creates a symlimks on it in lib/ directory of project. I am creating a patch when packing `gi-crystal` to make it work with symlinks. The main problem is that all derivations in nix store are read-only and changes in one derivation will change drv hash and all hashes, that depend on it. When we run `./bin/gi-crystal` it fails with error `Permission denied`, because it cannot write to read-only directory.

So, `gi-crystal` in nixpkgs works differently due to this nix features. All you need to do is add `gi-crystal` in build inputs, then run `gi-crystal`. It will scan `lib/` directory with all symlinks, then generate all bindings in `lib/gi-crystal` directory and copy self package to this.

Reply:
> Shortly, nix collects crystal dependencies in separated derivations, then creates a symlimks on it in lib/ directory of project. I am creating a patch when packing `gi-crystal` to make it work with symlinks. The main problem is that all derivations in nix store are read-only and changes in one derivation will change drv hash and all hashes, that depend on it. When we run `./bin/gi-crystal` it fails with error `Permission denied`, because it cannot write to read-only directory.

Jumping into the convo here... How about copying the `gi-crystal` derivation into the current rtfm directory and then use it ? That would fix both issues. You will be re-using the existing `gi-cristal` and it's directory is writable since you copied it into the `rtfm` build directory.

WDYT? Is it feasible?

Reply:
> How about copying the gi-crystal derivation into the current rtfm directory and then use it ?

This is exactly what I did when packing `gi-crystal`. But I actually automated this action so as not to do it constantly for different packages. 

I think that's better, don't you?

https://github.com/NixOS/nixpkgs/pull/243914/files#diff-1a3ec3be7dadadd498d82613b7eaba4d02f11f8470913c5cbdb7d54932c63954R53

Reply:
What I mean is to copy the output of `gi-crystal` into `rtfm` and then use it.
Once the compilation is over, you delete it and only copy the relevant `rtfm` files in `$out`.

WDYT?

Reply:
> What I mean is to copy the output of gi-crystal into rtfm and then use it.
> Once the compilation is over, you delete it and only copy the relevant rtfm files in $out.

I'm sorry but I don't understand what's the difference. If we want to make it work without patching `Makefile`, then we need to do several steps:
```bash
mkdir lib/gi-crystal
cp -r ${gi-crystal}/* lib/gi-crystal
mkdir bin/
mv lib/gi-crystal/bin/gi-crystal bin/
chmod +w lib/gi-crystal/src/auto
```

I think it is just unnecessary steps to be done on each package requires gi-crystal. 

And even if we do that, we still have to patch this to remove `shards install` command, because it will crash building due to `cannot fetch remote host github.com` something like this, I definitely don't remember what that error sounds like.

WDYT?

Reply:
I don't see the issue to do that all the time. I prefer that than patching the src.

But anyway, I was just passing by here thinking to provides new ideas, but I see that you already studied the question so I won't bother. I never used gi-crystal this is the reason why I might be wrong too.

Reply:
> I don't see the issue to do that all the time. I prefer that than patching the src.
> 
> But anyway, I was just passing by here thinking to provides new ideas, but I see that you already studied the question so I won't bother. I never used gi-crystal this is the reason why I might be wrong too.

Thanks for suggesting solutions to this problem. Patching sources was the reason why I didn't want to pack `gi-crystal`. My best idea is to make `gi-crystal` portable, this fits well with the nix ideology, but changes program behaviour.

Speaking of using gi-crystal, I understand the steps involved in implementing your idea. But others may not realize that you need to copy the `gi-crystal` directory, mv `gi-crystal` binary and change the write permissions.

Reply:
Ok... sad but ok.

It's weird that it needs to write in the `gi-crystal` directory.

Reply:
Looks like you guys are re-doing what shards already does, but at a package level. I mean... if derivations means a packaged code.